cmd/kubeadm/app/util/config/cluster_test.go

```diff
@@ -0,0 +1,222 @@
+							Secret: "abcdef0123456789",
+						},
+					},
+				},
+				NodeRegistration: kubeadm.NodeRegistrationOptions{
+					Name:      "node-foo",
+					CRISocket: "/var/run/custom-cri.sock",
+				},
+			},
+		},
+		{
+			name: "fetch invalid config from configMap",
+			testCfg: &kubeadm.MasterConfiguration{
+				KubernetesVersion: "v1.10.3",
+				API: kubeadm.API{
+					AdvertiseAddress: "1.2.3.4",
+					BindPort:         6443,
+				},
+				Etcd: kubeadm.Etcd{
+					Local: &kubeadm.LocalEtcd{
+						DataDir: "/some/path",
+					},
+				},
+				Networking: kubeadm.Networking{
+					ServiceSubnet: "10.96.0.1/12",
+					DNSDomain:     "cluster.local",
+					PodSubnet:     "10.0.1.15",
+				},
+				CertificatesDir: "/some/other/cert/dir",
+				BootstrapTokens: []kubeadm.BootstrapToken{
+					{
+						Token: &kubeadm.BootstrapTokenString{
+							ID:     "abcdef",
+							Secret: "abcdef0123456789",
+						},
+					},
+				},
+				NodeRegistration: kubeadm.NodeRegistrationOptions{
+					Name:      "node-foo",
+					CRISocket: "/var/run/custom-cri.sock",
+				},
+			},
+			expectErr: "couldn't parse subnet",
+		},
+		{
+			name:    "fetch valid config from cfgPath",
+			cfgPath: "testdata/defaulting/master/defaulted.yaml",
+			testCfg: &kubeadm.MasterConfiguration{
+				KubernetesVersion: "v1.10.3",
+				API: kubeadm.API{
+					AdvertiseAddress: "1.2.3.4",
+					BindPort:         6443,
+				},
+				Etcd: kubeadm.Etcd{
+					Local: &kubeadm.LocalEtcd{
+						DataDir: "/some/path",
+					},
+				},
+				Networking: kubeadm.Networking{
+					ServiceSubnet: "10.96.0.1/12",
+					DNSDomain:     "cluster.local",
+					PodSubnet:     "10.0.1.15",
+				},
+				CertificatesDir: "/some/other/cert/dir",
+				BootstrapTokens: []kubeadm.BootstrapToken{
+					{
+						Token: &kubeadm.BootstrapTokenString{
+							ID:     "abcdef",
+							Secret: "abcdef0123456789",
+						},
+					},
+				},
+				NodeRegistration: kubeadm.NodeRegistrationOptions{
+					Name:      "node-foo",
+					CRISocket: "/var/run/custom-cri.sock",
+				},
+			},
+			expectCfg: "testdata/defaulting/master/defaulted.yaml",
+		},
+		{
+			name:      "fetch invalid config from cfgPath",
+			cfgPath:   "testdata/validation/invalid_mastercfg.yaml",
+			expectErr: "was not of the form",
+		},
+		{
+			name:      "fetch config from not exist cfgPath",
+			cfgPath:   "testdata231/defaulting/master/defaulted.yaml",
+			expectErr: "no such file or directory",
+		},
+		{
+			name:      "fetch config when no configMap and no cfgPath",
+			expectErr: "not found",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			client := clientsetfake.NewSimpleClientset()
+			if tt.testCfg != nil {
+				err := uploadconfig.UploadConfiguration(tt.testCfg, client)
+				if err != nil {
+					t.Errorf("UploadConfiguration failed err: %v", err)
+				}
+			}
+			resultCfg, err := FetchConfigFromFileOrCluster(client, os.Stdout, "upgrade/config", tt.cfgPath)
+			if len(tt.expectErr) == 0 {
```

Comment:
instead of branching this whole block i would try to `t.Fatalf()` early and process the block after that.